src/service_daemon.rs

```diff
@@ -1430,12 +1437,48 @@ impl Zeroconf {
 
         buf.truncate(sz); // reduce potential processing errors
 
+        let is_unicast = !pktinfo.addr_dst.is_multicast();
+
+        // Ignore unicast packets outside the local link
+        if is_unicast {
+            let should_respond = match (pktinfo.addr_src.ip(), &intf_sock.intf.addr) {
+                (IpAddr::V4(src_ip), IfAddr::V4(intf)) => {
+                    if src_ip.is_loopback() {
+                        true
+                    } else {
+                        let src_ip: u32 = src_ip.into();
+                        let intf_ip: u32 = intf.ip.into();
+                        let intf_netmask: u32 = intf.netmask.into();
+                        // Is src_ip in local subnet?
+                        (intf_ip & intf_netmask) == (src_ip & intf_netmask)
+                    }
+                }
+                (IpAddr::V6(src_ip), &IfAddr::V6(_)) => {
+                    if src_ip.is_loopback() {
+                        true
+                    } else {
+                        // Does src_ip have on-link prefix?
+                        src_ip.segments()[0] & 0xffc0 == 0xfe80
+                    }
+                }
+                // Interface and source message IP versions do not match
+                _ => false,
+            };
+            if !should_respond {
+                return true;
+            }
+        };
+
         match DnsIncoming::new(buf) {
             Ok(msg) => {
                 if msg.is_query() {
-                    self.handle_query(msg, ip);
+                    self.handle_query(msg, ip, pktinfo.addr_src, is_unicast);
                 } else if msg.is_response() {
-                    self.handle_response(msg);
+                    if !is_unicast {
+                        self.handle_response(msg);
+                    } else {
+                        error!("Invalid message: unrequested unicast response");
```

Comment:
I don't fully understand this part. Does it mean we will not handle _any_ unicast responses?   

Is it because this lib does not have ability to send out query with `UNIQUE_RESPONSE` bit set yet?

Reply:
Yes exactly. Relevant RFC passage:

> A Multicast DNS querier MUST only accept unicast responses if
   they answer a recently sent query (e.g., sent within the last two
   seconds) that explicitly requested unicast responses.  A Multicast
   DNS querier MUST silently ignore all other unicast responses.

Therefore as the library currently will never send any queries requesting a unicast response, any unicast responses received will not be in answer to queries made by the library and therefore should be ignored. I'll add a comment to indicate this.